### ISSUE TITLE
Add ClouDNS Premium Support

### ISSUE THREAD

**Issue by mtalexan** [ddclient/ddclient#397]
ClouDNS has a simplistic interface for Dynamic DNS updates of specific records, but it's mostly intended for allowing implementation of Dynamic DNS without granting API access to free accounts.  It's a hashed API endpoint with no authentication that automatically takes the source IP address of the caller and sets it in the A record it was generated for (you use the ClouDNS GUI to generate the specific API link for the specific A record).  
This is currently what's implemented in ddclient, and the code notes that there's no documented way to verify what domain it goes to or whether the update succeeded, and no way to define the IP address it should be set to when called from behind a proxy.  

As a much better alternative, Premium (paid) accounts get access to a regular JSON API.  The API includes support for login authentication and zone-specific record changes.  An example of doing bulk updates of A records in zones is provided in PHP https://github.com/ClouDNS/cloudns-api-bulk-updates/blob/master/bulk-records-update/bulk-records-update.php.  

The interface needs to differentiate between an `auth-id` and a `sub-auth-id` since these require different POST parameters for which is being used, and the `sub-auth-id` is the mechanism for scoped access.  Notably updating the records is also slightly more complicated since it requires first querying the records of the proper type and hostname, then finding the assigned ID for that record so it can be modified.  Trying to simply add the record again generates a second record with a lower priority.

Here's a rough bash version of what it should do:
```bash
## Assumed to be available
# auth-id or sub-auth-id (must differentiate in input)
# password
# domain (the DNS zone being updated, must already exist)
# hostname (subdomain, e.g. "my.sub" when domain="example.com" will update "my.sub.example.com")
# ipv4addr (blank if not updating)
# ipv6addr (blank if not updating)
# ttl


BASE_URL="https://api.cloudns.net/dns"
ARGS=()
if [ -n "${auth-id}" ] ; then
  ARGS+=("--data-urlencode \"auth-id=${auth-id}\")
elif [ -n "${sub-auth-id}" ] ; then
  ARGS+=("--data-urlencode \"sub-auth-id=${sub-auth-id}\")
fi
ARGS+=("--data-urlencode \"auth-password=${password}\")

# set JSON_RESULT to the output
# arg1 is the name of what was being tested
check_json_result()
{
  local name=$1
  if [[ $(echo "${JSON_RESULT } | jq '.status'") == "Success" ] ; then
    echo "$name success"
    return 0
  else
    echo "$name error: $(echo "${JSON_RESULT } | jq '.statusDescription'")"
    return 1
  fi
}

# arg1 : IP address to set
# arg2 : A or AAAA (record type)
# Assumes ARGS contains:
#  host=${hostname}
#  domain=${domain}
#  auth-password=${password}
#  auth-id=${auth-id} or sub-auth-id=${sub-auth-id}
create_or_update_record()
{
  local addr=$1
  local type=$2

  # Does an record already exist?  Get the assigned ID if it does
  record_list=$(curl ${ARGS[@]} --data-urlencode "type=${type}" "${BASE_URL}/records.json")
  record_count=$(echo ${record_list} | jq ". | length")

  # we have to handle it as a string
  if $(expr "${record_count}" ">" "0") >/dev/null ; then
    # Update the existing record.

    # get the record ID to update and strip wrapping double quotes
    record_id=$(echo ${record_list} | jq ".[].id" | sed -e 's/"//g')
    
    mod_record_result=$(curl ${ARGS[@]} --data-urlencode "record-id=${record_id}" --data-urlencode "record=${addr}" --data-urlencode "ttl=${ttl}" "${BASE_URL}/mod-record.json")
    JSON_RESULT=${mod_record_result}
    check_json_result "Update"
  else
    # Add a new record

    add_record_result=$(curl ${ARGS[@]} --data-urlencode "record-type=${type}" --data-urlencode "record=${addr}" --data-urlencode "ttl=${ttl}" "${BASE_URL}/add-record.json")
    JSON_RESULT=${add_record_result}
    check_json_result "Create"
  fi
}

# Confirm the login works
login_test=$(curl ${ARGS[@]} "${BASE_URL}/login.json")
JSON_RESULT=${login_test}
if ! check_json_result "Login" ; then
  exit 1
fi

# Only after authentication check
ARGS+=("--data-urlencode \"domain=${domain}\"")
ARGS+=("--data-urlencode \"host=\"${hostname}")

if [ -n "${ipv4addr}" ] ; then
  create_or_update_record "${ipv4addr}" A
fi 

if [ -n "${ipv6addr}" ] ; then
  create_or_update_record "${ipv6addr}" AAAA
fi
```

**Comment by idallen** [comment]
The free ClouDNS does allow you to specify the IP you want (instead of using the IP address you're coming from) by appending `&proxy=1` to the API URL and setting the IP address you want in an `X-Forwarded-For` header.  Sample code:
~~~~~
#!/usr/bin/perl -w

use strict;

my $v4 = '10.1.2.3';

use HTTP::Tiny;
eval { HTTP::Tiny->can_ssl }
or die "Please ensure you have HTTP::Tiny (>= 0.055) and IO::Socket::SSL";

my $domain_A = 'some.domain.cloudns.nz';

# The xxxx... stuff is the API key you get from the ClouDNS GUI.
my $uri_A = 'https://ipv4.cloudns.net/api/dynamicURL/?q=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' . '&proxy=1';

my %hdr_A = (
    'X-Forwarded-For' => $v4
);

my %attr_A = (
    'default_headers' => \%hdr_A
);

my $res = HTTP::Tiny->new(%attr_A)->get($uri_A);
$res->{success} or die "$0: $hdr_A{'X-Forwarded-For'} $res->{status} $res->{reason}\n";
print {*STDERR} "$0: $hdr_A{'X-Forwarded-For'} notification sent successfully" if $ENV{DEBUG};

~~~~~
